### PR TITLE
Add "memory high" upscaling via cgroups

### PR DESCRIPTION
**Pull request by sharnoff**
High-level features added:

* agent: `/try-upscale` endpoint, taking [`api.MoreResources`](https://github.com/neondatabase/autoscaling/blob/246367af592737e526ba43fe5b85191cc62b10a0/pkg/api/types.go#L267-L272) — VM informant can request upscaling. Refer to the changes to ARCHITECTURE.md for a brief overview.
* informant: off-by-default cgroup memory.high event tracking — triggers `/try-upscale`.

Prior issues fixed:

* VM informant will deny downscaling if it's too close to triggering a memory.high event (fixes the "downscaling below current usage" issue).

Remaining tasks:

- [x] Add a little "allocate a bunch of memory" test program for the test VM
- [x] Actually test that it works (and fix the bugs) — with and without cgroup handling enabled.

Follow-up tasks:

- [ ] Integrate the cgroups stuff into `compute_ctl`

### CODE REVIEW

File: pkg/util/arith.go [hunk 1]
```diff
@@ -33,3 +33,31 @@ func Min[T constraints.Ordered](x, y T) T {
...
+	Swap(new I) (old I) //nolint:predeclared // same var names as methods
+}
+
+// AtomicMax atomically sets a to the maximum of *a and i, returning the old value at a.
+//
+// Eventually, one would *hope* that go could add atomic maximum (and minimum), so we don't need
+// this function (there are some ISAs that support such instructions). In the meantime though, this
+// tends to be pretty useful - event though it's not wait-free.
+func AtomicMax[A AtomicInt[I], I constraints.Integer](a A, i I) I {
+	for {
```
Comment (tychoish) [inline pkg/util/arith.go:54]: I _think_ that this is how `atomic.Value.CompareAndSwap()` is implemented, that not withstanding, I have this feeling that if you get into this situation you probably want a different sort of operation (e.g. just a mutex protecting an int, then you can let the runtime be smart (potentially) about letting the goroutine calling this function sleep until it can be woken up logically.)

File: vm_image/init [hunk 2]
```diff
@@ -5,6 +5,10 @@
 
 set -x
 
+# Setup the 'neon-test' cgroup:
+whoami
+mkdir /sys/fs/cgroup/neon-test
```
Comment (tychoish) [inline vm_image/init:10]: does this leak into a production environment? does the whoami need to stay?

File: pkg/informant/endpoints.go [hunk 3]
```diff
@@ -46,13 +113,64 @@ func (s *State) RegisterAgent(info *api.AgentDesc) (*api.InformantDesc, int, err
...
+		}
+
+		// Downscale:
+		//
+		// TODO: see similar note above. We shouldn't call methods on s.cgroup.mgr from here.
+		if err := s.cgroup.mgr.SetHighMem(newMemHigh); err != nil {
+			return false, "", fmt.Errorf("Error setting cgroup memory.high: %w", err)
+		}
+
+		status := fmt.Sprintf("Set cgroup memory.high down to %v MiB", float64(newMemHigh)/mib)
```
Comment (tychoish) [inline pkg/informant/endpoints.go:157]: %f for floats?

File: pkg/informant/endpoints.go [hunk 4]
```diff
@@ -3,22 +3,89 @@ package informant
...
+			//  1. Respond "ok" to a downscale request
+			//  2. Restart
+			//  3. Read system memory
+			//  4. Get downscaled (as approved earlier)
+			// A potential way to fix this would be writing to a file to record approved downscale
+			// operations.
+			if err := s.cgroup.setMemoryHigh(); err != nil {
+				return fmt.Errorf("Error setting initial cgroup memory.high: %w", err)
+			}
+			go s.cgroup.handleCgroupSignalsLoop(config)
```
Comment (tychoish) [inline pkg/informant/endpoints.go:86]: one of my "rules" is "if you start a go routine without passing it a context, you should be really sure this is right, and then probably not do it," there are exceptions, but I've always regretted making them, for anything non-trivial.

File: pkg/informant/cgroupmanager.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	memEvents, eventErrCh := manager.EventChan()
+	sendEvent, recvEvent := util.NewCondChannelPair()
+
+	highEventCount := &atomic.Uint64{}
+	errCh := make(chan error, 1)
+
+	// Long-running handler task for memory events
+	go func() {
+		for {
+			select {
```
Comment (tychoish) [inline pkg/informant/cgroupmanager.go:61]: how does this daemon thread handle shutdown

File: pkg/informant/cgroup.go [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+// handleMemoryHighEvent performs the "freeze cgroup, request upscale, thaw cgroup" operation, in
+// response to a "memory high" event for the cgroup
+//
+// This method waits on s.agents.UpscaleEvents(), so incorrect behavior will occur if it's called at
+// the same time as anything else that waits on the upscale events. For that reason, both this
+// function and s.setMemoryHigh() are dispatched from within s.handleCgroupSignalsLoop().
+//
+// This method MUST be called while holding s.updateLock.
+func (s *CgroupState) handleMemoryHighEvent(config CgroupConfig) error {
+	klog.Infof("Received memory high event. Freezing cgroup")
```
Comment (tychoish) [inline pkg/informant/cgroup.go:160]: is it meaingful to tell us anything about that event?

File: pkg/informant/cgroup.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+	// Immediately freeze the cgroup before doing anything else.
+	if err := s.mgr.Freeze(); err != nil {
+		return fmt.Errorf("Error freezing cgroup: %w", err)
+	}
+
+	startTime := time.Now()
+
+	// Start a timer for the maximum time we'll leave the cgroup frozen for:
+	maxWaitBeforeThaw := time.Millisecond * time.Duration(config.MaxUpscaleWaitMillis)
```
Comment (tychoish) [inline pkg/informant/cgroup.go:170]: I always prefer "num * <unit>" also because duration is just an int64 if your config variable is in int64 you *might* not need to convert. 

also I'd probably use `MS` over `Millis`

File: pkg/informant/cgroup.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+			select {
+			case <-s.upscaleEventsRecvr.Recv():
+				klog.Warningf("Received memory high event, but also upscale event. Handling upscale and ignoring memory high")
+				func() {
+					s.updateLock.Lock()
+					defer s.updateLock.Unlock()
+
+					if err := s.setMemoryHigh(); err != nil {
+						panic(fmt.Errorf("Error handling upscale signal: %w", err))
+					}
```
Comment (tychoish) [inline pkg/informant/cgroup.go:134]: maybe make these methods? 

all of the panics worry me? I think if you expect that all of these errors are unrecoverable by their nature, I'd just panic in the functions that produce them rather than go through the effort of returning errors that are just going to convert them to panics? 

(unless these errors represent invariant violations, panic'ing this much seems to be indicative of some sort of other design flaw)

File: pkg/informant/agent.go [hunk 9]
```diff
@@ -702,3 +764,45 @@ func (a *Agent) Resume(timeout time.Duration) error {
...
+// SpawnRequestUpscale requests that the Agent increase the resource allocation to this VM
+func (a *Agent) SpawnRequestUpscale(timeout time.Duration, handleError func(error)) {
+	// Quick unregistered check
+	select {
+	case <-a.unregistered.Recv():
+		klog.Warningf(
+			"RequestUpscale called for Agent %s/%s that is already unregistered (probably *not* a race?)",
+			a.serverAddr, a.id,
+		)
+		handleError(context.Canceled)
```
Comment (tychoish) [inline pkg/informant/agent.go:777]: this seems like a weird sentinel error to pass around, but sure.

File: pkg/informant/agent.go [hunk 10]
```diff
@@ -321,6 +344,45 @@ func (s *AgentSet) RegisterNewAgent(info *api.AgentDesc) (uint32, int, error) {
...
+	defer s.lock.Unlock()
+
+	s.wantsMemoryUpscale = true
+
+	// explicitly copy s.current so that our error handling points to the right Agent
+	if agent := s.current; agent != nil {
+		agent.SpawnRequestUpscale(AgentUpscaleTimeout, func(err error) {
+			if errors.Is(err, context.Canceled) {
+				return
+			}
```
Comment (tychoish) [inline pkg/informant/agent.go:366]: what about deadline exceeded?

File: pkg/api/types.go [hunk 11]
```diff
@@ -262,15 +270,52 @@ type UnregisterAgent struct {
...
+// MoreResources holds the data associated with a MoreResourcesRequest
 type MoreResources struct {
 	// Cpu is true if the VM informant is requesting more CPU
 	Cpu bool `json:"cpu"`
 	// Memory is true if the VM informant is requesting more memory
 	Memory bool `json:"memory"`
 }
 
+// Not returns the field-wise logical "not" of m
+func (m MoreResources) Not() MoreResources {
```
Comment (tychoish) [inline pkg/api/types.go:291]: none of these methods are actually called, I think?

File: pkg/informant/cgroupmanager.go [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+func NewCgroupManager(groupName string) (*CgroupManager, error) {
+	mode := cgroups.Mode()
+	if mode != cgroups.Unified && mode != cgroups.Legacy {
+		var modeString string
+		switch mode {
+		case cgroups.Unavailable:
+			modeString = "Unavailable"
+		case cgroups.Legacy:
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/informant/cgroupmanager.go:35]: Ah yep, good catch. The condition above should be `... && mode != cgroups.Hybrid`, not `cgroups.Legacy`.

File: pkg/informant/cgroupmanager.go (hunk 5 above)
Comment (sharnoff) [reply pkg/informant/cgroupmanager.go:61]: Refer to the comment 10 lines up:

https://github.com/neondatabase/autoscaling/blob/23e7b4c5ff02496946521d7540695e8d09b562c1/pkg/informant/cgroupmanager.go#L49-L51

And

#12:

> - [ ] Cleanup on VM informant exit
>     - note: Currently (2023-01-10), NeonVM doesn't support clean shutdown, so this is not immediately applicable

tl;dr: Currently nothing else _can_ shut down, so there isn't anything it _can_ do.

File: pkg/informant/cgroupmanager.go [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+	// Long-running handler task for memory events
+	go func() {
+		for {
+			select {
+			case event := <-memEvents:
+				highCount := event.High
+				oldHighCount := util.AtomicMax(highEventCount, highCount)
+				if highCount > oldHighCount {
+					sendEvent.Send()
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/informant/cgroupmanager.go:66]: `util.AtomicMax` isn't wait-free, but there's only two threads that can access `highEventCount` at that point.

`(*util.CondChannelSender).Send()` is wait-free.

File: pkg/informant/cgroupmanager.go [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+				}
+			case err, ok := <-eventErrCh:
+				if err == nil && !ok {
+					errCh <- errors.New("Memory event channel closed without error")
+				} else {
+					errCh <- fmt.Errorf("Error while waiting for memory events: %w", err)
+				}
+				return
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/informant/cgroupmanager.go:74]: There aren't any other threads. `CgroupManager` stores it as a `<-chan error`.

File: pkg/informant/cgroupmanager.go [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+	}, nil
+}
+
+// TODO: no way to do this with github.com/containerd/cgroups ? Seems like that should be
+// exposed to the user... We *can* just parse it directly, but it's a bit annoying.
+func parseMemoryEvents(groupName string) (*cgroup2.Event, error) {
+	path := cgroupPath(groupName, "memory.events")
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading file at %q: %w", path, err)
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/informant/cgroupmanager.go:102]: Can you elaborate?

File: pkg/informant/cgroupmanager.go [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+	}
+
+	event := cgroup2.Event{}
+	valueMap := map[string]struct {
+		v   *uint64
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/informant/cgroupmanager.go:107]: Not sure what you're saying

File: pkg/informant/endpoints.go (hunk 3 above)
Comment (sharnoff) [reply pkg/informant/endpoints.go:157]: %v because it doesn't include decimals if it doesn't need to, which is _almost_ guaranteed to be the case.

For floats, %v is equivalent to %g - might make sense to use that instead tbh.

File: pkg/util/arith.go [hunk 17]
```diff
@@ -33,3 +33,31 @@ func Min[T constraints.Ordered](x, y T) T {
 		return y
 	}
 }
+
+// AtomicInt represents the shared interface provided by various atomic.<NAME> integers
+//
+// This interface type is primarily used by AtomicMax.
+type AtomicInt[I any] interface {
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/util/arith.go:40]: What do you mean?

File: pkg/util/arith.go (hunk 1 above)
Comment (sharnoff) [reply pkg/util/arith.go:54]: > I think that this is how atomic.Value.CompareAndSwap() is implemented

_Kind of_, but not quite imo. It's [mostly pretty readable](https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/sync/atomic/value.go;l=135), though!

File: vm_image/init (hunk 2 above)
Comment (sharnoff) [reply vm_image/init:10]: good catch - the `whoami` is just for debugging. I'll make a note to remove it once I'm done testing stuff.

File: cmd/vm-informant/main.go [hunk 18]
```diff
@@ -15,8 +16,32 @@ func main() {
 	klog.Infof("buildInfo.GitInfo:   %s", buildInfo.GitInfo)
 	klog.Infof("buildInfo.GoVersion: %s", buildInfo.GoVersion)
 
+	nullString := "\x00"
+	var cgroupName string
+	flag.StringVar(&cgroupName, "cgroup", nullString, "Sets the cgroup to monitor (optional)")
```
[the remark this answers is not in the document]
Comment (tychoish) [reply cmd/vm-informant/main.go:21]: but _should_ it be a cgroup? would we want people to do this? seems like we can be slightly more restrictive than the kernel, to reduce confusion in general.

File: pkg/agent/runner.go [hunk 19]
```diff
@@ -1476,6 +1500,28 @@ func (s *atomicUpdateState) computeUnitsBounds() (uint16, uint16) {
...
+// This method does not respect any bounds on Compute Units placed by the VM's maximum or minimum
+// resource allocation.
+func (s *atomicUpdateState) requiredCUForRequestedUpscaling() uint16 {
+	var required uint16
+
+	// note: floor(x / M) + 1 gives the minimum integer value greater than x / M.
+
+	if s.requestedUpscale.Cpu {
+		required = util.Max(required, s.vm.Cpu.Use/s.computeUnit.VCPU+1)
+	}
```
[the remark this answers is not in the document]
Comment (tychoish) [reply pkg/agent/runner.go:1517]: maybe? I definitely had to read it more than times, and cross reference to the definition of `util.Max` which spins over an atomic. I'd feel a bit better about this if the util functions had unittests. 

I think I had to read it more than once because doing a max of two numbers when one is obviously a zero makes me think something fishy is up, or looks like it's unintentional.

File: pkg/informant/agent.go [hunk 20]
```diff
@@ -321,6 +344,45 @@ func (s *AgentSet) RegisterNewAgent(info *api.AgentDesc) (uint32, int, error) {
...
+	// FIXME: we should assign a timeout to these upscale requests, so that we don't continue trying
+	// to upscale after the demand has gone away.
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.wantsMemoryUpscale = true
+
+	// explicitly copy s.current so that our error handling points to the right Agent
+	if agent := s.current; agent != nil {
```
[the remark this answers is not in the document]
Comment (tychoish) [reply pkg/informant/agent.go:362]: I've seen a convention in struct delcaration that says "if you're not going to mutex every field in the struct, put the mutex and the covered fields near the end in a group so it's clear what they protect." which I certainly haven't always done myself, but it definitely makes that more clear. 

One of the problems is that you only get immutability in go structs by copying the data (which all this being pointers doesn't do) or by pinky swearing that no code is going to modify a value.  

While I'm convinced that this is correct, this code *looks* like a bug, or is "valid form using the idiom of a common bug", so even though it's correct, the look and comment sort of makes it seem more suspect?

File: pkg/informant/cgroupmanager.go [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+
+	cgroups "github.com/containerd/cgroups/v3"
+	"github.com/containerd/cgroups/v3/cgroup2"
+
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+type CgroupManager struct {
+	MemoryHighEvent util.CondChannelReceiver
+	ErrCh           <-chan error
```
[the remark this answers is not in the document]
Comment (tychoish) [reply pkg/informant/cgroupmanager.go:22]: when I've used error channels in the past, I tend to make them buffered to the same extent as the number of threads/possible senders, or put the consume of the thread *really* close to the definition.

File: pkg/informant/cgroupmanager.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (tychoish) [reply pkg/informant/cgroupmanager.go:102]: oh, just if you're writing code, you have an object of type `error` you know that it represents a conceptual error, the output is almost always `ERROR: <message>` so putting the word "error" in the text of the error, is a lot.

File: pkg/informant/cgroupmanager.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (tychoish) [reply pkg/informant/cgroupmanager.go:66]: got it... 

I think this is, then, safely in the "this reads a bit suspect if you don't have the full context for how this works and a good mental model for the go routines in the system" category.

File: pkg/informant/cgroupmanager.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (tychoish) [reply pkg/informant/cgroupmanager.go:107]: Let me attempt a different way: 

- why isn't the structure of the value type of this map named?
- why do you have a boolean field in the value of a map, for which accessing values already has an "ok" boolean value to let you now if its set?
- what is the meaning of:
  - `valueMap["foo"] = struct{v: nil, set: true}`

File: pkg/util/arith.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (tychoish) [reply pkg/util/arith.go:40]: I believe that this is valid: `name := AtomicInt[string]`, because `I` can be any type, not just an integer, which is confusing.

File: pkg/informant/agent.go (hunk 9 above)
Comment (tychoish) [reply pkg/informant/agent.go:777]: just declare a sentinel error in this named of `ErrRequestConflictTimeout` or something similar.

File: pkg/agent/runner.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/agent/runner.go:1521]: > `util.Max` which spins over an atomic

I think you're confusing it with `util.AtomicMax`? The definition of `util.Max` is:

https://github.com/neondatabase/autoscaling/blob/bfbf9873c8ea8400536b1b54999c0b421ac48687/pkg/util/arith.go#L19-L26

> doing a max of two numbers when one is obviously a zero makes me think something fishy is up

Fair enough. I think it's personal taste — for me, reading it without a maximum on the first means I have to double-check that the high-level logic on the first & second are still fundamentally the same, even though they're phrased differently.

**Comment by sharnoff** [comment]
Pushed some updates. To get it working, opened https://github.com/neondatabase/neonvm/pull/33 — blocked on that before merging.

**Comment by sharnoff** [comment]
Spent a while dealing with issues from an over-eager OOM killer after memory hotplug failure, but it seems like fixing some related VM informant bugs stopped that from happening - even though they couldn't possibly be the *cause*.

File: pkg/api/types.go (hunk 11 above)
Comment (sharnoff) [reply pkg/api/types.go:291]: Removed `Or`.

File: pkg/informant/cgroup.go (hunk 7 above)
Comment (sharnoff) [reply pkg/informant/cgroup.go:177]: Opened #35.

File: pkg/util/arith.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/util/arith.go:40]: `AtomicInt` is an interface, so you can't do that directly. You could 100% declare a type like that, but probably will run into trouble constructing it (because realistically nothing will have the appropriate methods returning `string`s)

File: vm_image/init (hunk 2 above)
Comment (sharnoff) [reply vm_image/init]: Done

File: pkg/informant/endpoints.go (hunk 3 above)
Comment (sharnoff) [reply pkg/informant/endpoints.go:161]: Changed to %g.

File: pkg/informant/cgroupmanager.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/informant/cgroupmanager.go]: One of the recent commits expanded on a comment here, to provide some more context.

File: pkg/util/arith.go [hunk 22]
```diff
@@ -33,3 +33,31 @@ func Min[T constraints.Ordered](x, y T) T {
...
+	Add(delta I) (new I)                      //nolint:predeclared // same var names as methods
+	CompareAndSwap(old, new I) (swapped bool) //nolint:predeclared // same var names as methods
+	Load() I
+	Store(val I)
+	Swap(new I) (old I) //nolint:predeclared // same var names as methods
+}
+
+// AtomicMax atomically sets a to the maximum of *a and i, returning the old value at a.
+//
+// Eventually, one would *hope* that go could add atomic maximum (and minimum), so we don't need
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/util/arith.go]: Fixed

**Comment by sharnoff** [comment]
Ok, think this is good to merge. I'll give it a once-over before merging tomorrow, with a follow-up PR to https://github.com/neondatabase/neon to add cgroup handling there as well.

This feature has some unfortunate interactions with other open issues, particularly:

1. Our current metrics and scaling algorithm mean that it's hard to scale back down after increasing due to `/try-upscale` requests (if it can't go all the way back down to the level indicated by load average, it won't go down at all)
2. Some of the raciness around NeonVM means we have to "trust" what the autoscaler-agent says when it calls the `/upscale` endpoint (see: https://github.com/neondatabase/autoscaling/issues/23#issuecomment-1398805853, point (2))

I believe both of these require protocol changes, alongside some of the other changes from #27.
